### PR TITLE
Device requirements

### PR DESCRIPTION
**Pull request by nicelhc13**
to give an idea. wip.

### CODE REVIEW

File: src/python/parla/common/spawn.py [hunk 1]
```diff
@@ -39,7 +42,11 @@ def closure():
 
 
 # @profile
-def spawn(task=None,  dependencies=[], vcus=1):
+def spawn(task=None,
+          dependencies=[],
+          # TODO(hc): Do we support TaskID? (IIRC, it will be removed?)
```
Comment (dialecticDolt) [inline src/python/parla/common/spawn.py:47]: It's already removed. TaskSpaces return a ComputeTask. If the task argument is not a ComputeTask already, it must be a hashable type and we put it in the global taskspace.

File: src/c/backend/task.cpp [hunk 2]
```diff
@@ -301,3 +305,54 @@ void InnerTask::set_complete() {
...
+  if (req_addition_mode_ == SingleDevAdd) {
+    dev_res_reqs_.AppendDeviceRequirementOption(dev_req);
+  } else if (req_addition_mode_ % 2 == 0) { /* Architecture requirement */
+    tmp_arch_req_->AppendDeviceRequirementOption(dev_req);
+  } else if (req_addition_mode_ == MultiDevAdd) {
+    tmp_multdev_reqs_->AppendDeviceRequirement(dev_req);
+  }
+}
+
+void InnerTask::begin_arch_req_addition() {
```
Comment (dialecticDolt) [inline src/c/backend/task.cpp:325]: What's the benefit of this pattern (just curious)?

**Comment by dialecticDolt** [comment]
Can you make the CuPy dependency in device manager optional (so we can run tests on machines without gpus)?

File: src/c/backend/task.cpp (hunk 2 above)
Comment (nicelhc13) [reply src/c/backend/task.cpp:325]: Basically this pattern was came from the lack of idea. You can understand this as a transaction-based method. We need to handle nested/recursive patterns even though this is at most 2 levels. I have hierarchical vectors/classes and the lower level vectors containing requirements should be pushed to the higher level. For example, architecture requirement consisting of multiple device requirements could be pushed into multi-device requirement vector or the topmost vector. I can distinguish the levels easily with those functions.

**Comment by nicelhc13** [comment]
> Can you make the CuPy dependency in device manager optional (so we can run tests on machines without gpus)?
> 
> Edit: Actually, I needed to patch it already :) so I'll just push this as part of my merge.

I think device_manager already did it, didn't it?

**Review by dialecticDolt — APPROVED** [review]

**Pull request merged**